File: python/mirage/commands/builtin/linear/linear_issue_update.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import json

from mirage.accessor.linear import LinearAccessor
from mirage.commands.builtin.linear._input import resolve_text_input
from mirage.commands.registry import command
from mirage.commands.spec.types import (CommandSpec, FlagView, OperandKind,
                                        Option)
from mirage.core.linear._client import issue_update, resolve_issue_id
from mirage.core.linear.normalize import normalize_issue
from mirage.io.stream import yield_bytes
from mirage.io.types import ByteSource, IOResult
from mirage.types import PathSpec

SPEC = CommandSpec(options=(
    Option(long="--issue_id", value_kind=OperandKind.TEXT),
    Option(long="--issue_key", value_kind=OperandKind.TEXT),
    Option(long="--title", value_kind=OperandKind.TEXT),
    Option(long="--description", value_kind=OperandKind.TEXT),
    Option(long="--description_file", value_kind=OperandKind.PATH),
), )


@command("linear issue update", resource="linear", spec=SPEC)
async def linear_issue_update(
    accessor: LinearAccessor,
    paths: list[PathSpec],
    *texts: str,
    stdin: ByteSource | None = None,
    **_extra: object,
) -> tuple[ByteSource | None, IOResult]:
    fl = FlagView(_extra, spec=SPEC)
    config = accessor.config
    issue_id = await resolve_issue_id(
        config,
        issue_id=fl.as_str("issue_id"),
        issue_key=fl.as_str("issue_key"),
    )
    title = fl.as_str("title")
    description = None
    if (_extra.get("description") is not None
            or _extra.get("description_file") is not None
            or stdin is not None):
        description = await resolve_text_input(
            config,
            inline_text=fl.as_str("description"),
            file_path=fl.as_str("description_file"),
            stdin=stdin,
            error_message="description is required",
        )
    issue = await issue_update(config,
                               issue_id=issue_id,
                               title=title,
                               description=description)
    return yield_bytes(
        json.dumps(normalize_issue(issue),
                   ensure_ascii=False,
                   separators=(",", ":")).encode()), IOResult()

File: typescript/packages/core/src/commands/builtin/linear/linear_issue_update.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { LinearAccessor } from '../../../accessor/linear.ts'
import { issueUpdate, resolveIssueId } from '../../../core/linear/_client.ts'
import { normalizeIssue } from '../../../core/linear/normalize.ts'
import { IOResult } from '../../../io/types.ts'
import { ResourceName, type PathSpec } from '../../../types.ts'
import { command, type CommandFnResult, type CommandOpts } from '../../config.ts'
import { CommandSpec, OperandKind, Option } from '../../spec/types.ts'
import { resolveTextInput } from './_input.ts'

const ENC = new TextEncoder()

const SPEC = new CommandSpec({
  options: [
    new Option({ long: '--issue_id', valueKind: OperandKind.TEXT }),
    new Option({ long: '--issue_key', valueKind: OperandKind.TEXT }),
    new Option({ long: '--title', valueKind: OperandKind.TEXT }),
    new Option({ long: '--description', valueKind: OperandKind.TEXT }),
    new Option({ long: '--description_file', valueKind: OperandKind.PATH }),
  ],
})

async function linearIssueUpdateCommand(
  accessor: LinearAccessor,
  _paths: PathSpec[],
  _texts: string[],
  opts: CommandOpts,
): Promise<CommandFnResult> {
  const issueId = await resolveIssueId(
    accessor.transport,
    typeof opts.flags.issue_id === 'string' ? opts.flags.issue_id : null,
    typeof opts.flags.issue_key === 'string' ? opts.flags.issue_key : null,
  )
  const title = typeof opts.flags.title === 'string' ? opts.flags.title : null
  const inlineDesc = typeof opts.flags.description === 'string' ? opts.flags.description : null
  const descFile =
    typeof opts.flags.description_file === 'string' ? opts.flags.description_file : null
  let description: string | undefined
  if (inlineDesc !== null || descFile !== null || opts.stdin !== null) {
    description = await resolveTextInput(accessor.transport, {
      inlineText: inlineDesc,
      filePath: descFile,
      stdin: opts.stdin,
      errorMessage: 'description is required',
    })
  }
  const issue = await issueUpdate(accessor.transport, {
    issueId,
    ...(title !== null ? { title } : {}),
    ...(description !== undefined ? { description } : {}),
  })
  return [ENC.encode(JSON.stringify(normalizeIssue(issue))), new IOResult()]
}

export const LINEAR_ISSUE_UPDATE = command({
  name: 'linear issue update',
  resource: ResourceName.LINEAR,
  spec: SPEC,
  fn: linearIssueUpdateCommand,
  write: true,
})
